File: src/lib/workouts/fallback-generator.ts
import type { GeneratedWorkoutPlan, WorkoutExercise, WorkoutProfileSnapshot } from "@/lib/workouts/types";

const strengthPool: WorkoutExercise[] = [
  { name: "Goblet squat", muscle_group: "Lower body", equipment: "dumbbells", sets: 3, reps: "8-10", rest_seconds: 75, instructions: ["Hold one weight at chest height.", "Sit hips back and keep knees tracking over toes.", "Drive through the floor to stand tall."] },
  { name: "Push-up", muscle_group: "Chest", equipment: "bodyweight", sets: 3, reps: "6-12", rest_seconds: 60, instructions: ["Set hands under shoulders and brace your core.", "Lower as one unit until chest nears the floor.", "Press back up without flaring elbows wide."] },
  { name: "Bent-over row", muscle_group: "Back", equipment: "dumbbells", sets: 3, reps: "10-12", rest_seconds: 60, instructions: ["Hinge at the hips with a long spine.", "Pull weights toward your lower ribs.", "Control the lowering phase."] },
  { name: "Reverse lunge", muscle_group: "Lower body", equipment: "bodyweight", sets: 2, reps: "8 each side", rest_seconds: 60, instructions: ["Step one foot back into a split stance.", "Lower until both knees bend comfortably.", "Push through the front foot to return."] },
  { name: "Plank", muscle_group: "Core", equipment: "bodyweight", sets: 3, reps: "30-45 sec", rest_seconds: 45, instructions: ["Stack elbows under shoulders.", "Squeeze glutes and keep ribs tucked.", "Breathe steadily without sagging."] },
];

const endurancePool: WorkoutExercise[] = [
  { name: "Marching high knees", muscle_group: "Conditioning", equipment: "bodyweight", sets: 3, reps: "45 sec", rest_seconds: 30, instructions: ["Stand tall and alternate knee drives.", "Pump arms naturally.", "Keep the pace brisk but controlled."] },
  { name: "Step-back squat", muscle_group: "Lower body", equipment: "bodyweight", sets: 3, reps: "12-15", rest_seconds: 45, instructions: ["Squat to a comfortable depth.", "Step one foot back after each rep.", "Alternate sides while keeping the chest proud."] },
  { name: "Mountain climber", muscle_group: "Core", equipment: "bodyweight", sets: 3, reps: "30 sec", rest_seconds: 30, instructions: ["Start in a strong plank.", "Drive knees toward the chest one at a time.", "Keep hips steady."] },
  { name: "Glute bridge", muscle_group: "Posterior chain", equipment: "bodyweight", sets: 3, reps: "12-15", rest_seconds: 45, instructions: ["Lie on your back with feet planted.", "Lift hips by squeezing glutes.", "Pause at the top, then lower slowly."] },
];

function normalized(value: string | null | undefined) {
  return (value ?? "").toLowerCase();
}

function allowedExercise(exercise: WorkoutExercise, equipment: string[]) {
  const owned = equipment.map((item) => item.toLowerCase());
  return exercise.equipment === "bodyweight" || owned.some((item) => exercise.equipment.includes(item) || item.includes(exercise.equipment));
}

function pickExercises(profile: WorkoutProfileSnapshot) {
  const goal = normalized(profile.goal);
  const base = goal.includes("endurance") || goal.includes("fat") ? endurancePool : strengthPool;
  const available = base.filter((exercise) => allowedExercise(exercise, profile.equipment));
  const selected = (available.length >= 4 ? available : [...available, ...base.filter((exercise) => exercise.equipment === "bodyweight")]).slice(0, 5);
  return selected.length ? selected : strengthPool.filter((exercise) => exercise.equipment === "bodyweight").slice(0, 4);
}

function difficulty(profile: WorkoutProfileSnapshot) {
  const experience = normalized(profile.experience_level || profile.fitness_level);
  if (experience.includes("experienced") || experience.includes("athlete") || experience.includes("high")) return "advanced";
  if (experience.includes("intermediate") || experience.includes("moderate")) return "intermediate";
  return "beginner";
}

export function generateFallbackWorkout(profile: WorkoutProfileSnapshot): GeneratedWorkoutPlan {
  const minutes = profile.available_minutes ?? 35;
  const exercises = pickExercises(profile).map((exercise) => ({
    ...exercise,
    sets: difficulty(profile) === "beginner" ? Math.min(exercise.sets, 2) : exercise.sets,
  }));

  return {
    title: `${profile.goal ?? "Personalized"} session`,
    summary: `A ${minutes}-minute workout tailored to your goal, level, equipment, and schedule.`,
    duration_minutes: minutes,
    difficulty: difficulty(profile),
    goal: profile.goal ?? "General wellness",
    warmup: ["3 minutes easy cardio or marching in place", "10 bodyweight good mornings", "10 arm circles each direction"],
    blocks: [{ name: "Main training block", focus: profile.goal ?? "Full body fitness", exercises }],
    cooldown: ["Slow nasal breathing for 60 seconds", "Standing quad stretch", "Chest and lat stretch"],
    coaching_notes: [
      "Move with clean form before adding speed or load.",
      profile.limitations ? `Respect this limitation: ${profile.limitations}` : "Stop if pain appears and choose a gentler range of motion.",
      "Log how the session felt so future workouts can adapt.",
    ],
  };
}

File: services/ai/fallback_generator.py
"""Deterministic workout generator.

This is the safety net behind Gemini. It never calls the network, so the
product stays usable when the API key is missing, the quota is exhausted, or
the model returns something that fails validation.

It draws from the same exercise library the model is given, so plans produced
by either path reference real library entries and keep their demos and
live-form-tracking capability.
"""

from __future__ import annotations

from typing import Any

# Used only when the caller supplies no library at all.
_BUILTIN_LIBRARY: list[dict[str, Any]] = [
    {"slug": "bodyweight-squat", "name": "Bodyweight squat", "muscle_group": "Lower body",
     "equipment": "bodyweight", "difficulty": "beginner", "vision_kind": "squat",
     "instructions": ["Stand with feet shoulder width apart.", "Sit your hips back and down.",
                      "Descend until your thighs are parallel.", "Drive through your whole foot to stand."]},
    {"slug": "push-up", "name": "Push-up", "muscle_group": "Chest",
     "equipment": "bodyweight", "difficulty": "beginner", "vision_kind": "pushup",
     "instructions": ["Set hands slightly wider than your shoulders.", "Brace into a straight plank.",
                      "Lower until your chest nears the floor.", "Press back up without sagging."]},
    {"slug": "reverse-lunge", "name": "Reverse lunge", "muscle_group": "Lower body",
     "equipment": "bodyweight", "difficulty": "beginner", "vision_kind": "lunge",
     "instructions": ["Stand tall, feet hip width apart.", "Step one foot back and lower.",
                      "Keep your torso upright.", "Push through the front foot to return."]},
    {"slug": "glute-bridge", "name": "Glute bridge", "muscle_group": "Glutes",
     "equipment": "bodyweight", "difficulty": "beginner", "vision_kind": "glute_bridge",
     "instructions": ["Lie on your back with knees bent.", "Squeeze your glutes and lift your hips.",
                      "Pause at the top.", "Lower with control."]},
    {"slug": "plank", "name": "Plank", "muscle_group": "Core",
     "equipment": "bodyweight", "difficulty": "beginner", "vision_kind": None,
     "instructions": ["Stack elbows under shoulders.", "Form one straight line.",
                      "Tuck your ribs and squeeze your glutes.", "Breathe steadily."]},
    {"slug": "marching-high-knees", "name": "Marching high knees", "muscle_group": "Conditioning",
     "equipment": "bodyweight", "difficulty": "beginner", "vision_kind": None,
     "instructions": ["Stand tall.", "Drive knees up one at a time.", "Keep a brisk rhythm.",
                      "Land softly."]},
]

# Default programming per muscle group, keyed by goal family.
_REP_SCHEMES = {
    "strength": {"sets": 4, "reps": "5-6", "rest": 120},
    "muscle": {"sets": 3, "reps": "8-12", "rest": 75},
    "endurance": {"sets": 3, "reps": "15-20", "rest": 45},
    "general": {"sets": 3, "reps": "10-12", "rest": 60},
}

_MUSCLE_ORDER = ["Lower body", "Chest", "Back", "Shoulders", "Glutes", "Hamstrings", "Core", "Conditioning"]


def _goal_family(goal: str | None) -> str:
    text = (goal or "").lower()
    if "strength" in text:
        return "strength"
    if "muscle" in text or "build" in text:
        return "muscle"
    if "endurance" in text or "fat" in text or "lose" in text:
        return "endurance"
    return "general"


def _base_difficulty(profile: dict[str, Any]) -> str:
    level = f"{profile.get('experience_level') or ''} {profile.get('fitness_level') or ''}".lower()
    if any(token in level for token in ("experienced", "athlete", "high")):
        return "advanced"
    if any(token in level for token in ("intermediate", "moderate")):
        return "intermediate"
    return "beginner"


def resolve_difficulty(profile: dict[str, Any], context: dict[str, Any]) -> str:
    """Adjusts the base difficulty using recent completion and form data."""
    base = _base_difficulty(profile)

    sessions = context.get("recent_sessions") or []
    if not sessions:
        return base

    expected = 0
    completed = 0
    form_scores: list[float] = []
    for session in sessions[:3]:
        data = session.get("session_data") or {}
        if data.get("total_sets"):
            expected += int(data.get("total_sets") or 0)
            completed += int(data.get("completed_sets") or 0)
        if data.get("form_score") is not None:
            form_scores.append(float(data["form_score"]))

    completion = completed / expected if expected else 1.0
    average_form = sum(form_scores) / len(form_scores) if form_scores else 80.0

    if completion >= 0.95 and average_form >= 85:
        return f"{base}+"
    if completion < 0.60:
        return f"{base}-" if base != "beginner" else base
    return base


def _available(entry: dict[str, Any], equipment: list[str]) -> bool:
    required = str(entry.get("equipment", "bodyweight")).lower()
    if required == "bodyweight":
        return True
    owned = [str(item).lower() for item in equipment]
    return any(required in item or item in required for item in owned)


def _avoids_limitations(entry: dict[str, Any], limitations: str | None) -> bool:
    """Conservative keyword screen so stated injuries are respected."""
    if not limitations:
        return True
    text = limitations.lower()
    muscle = str(entry.get("muscle_group", "")).lower()
    slug = str(entry.get("slug", "")).lower()

    screens = [
        (("knee",), ("lower body", "glutes")),
        (("shoulder",), ("shoulders", "chest")),
        (("back", "spine", "disc"), ("back", "hamstrings")),
        (("wrist",), ("chest",)),
    ]
    for keywords, blocked_groups in screens:
        if any(keyword in text for keyword in keywords):
            if muscle in blocked_groups:
                return False
            if "squat" in slug or "lunge" in slug:
                if any(keyword in text for keyword in ("knee",)):
                    return False
    return True


def generate_workout(profile: dict[str, Any], context: dict[str, Any]) -> dict[str, Any]:
    library = context.get("exercise_library") or _BUILTIN_LIBRARY
    equipment = profile.get("equipment") or []
    limitations = profile.get("limitations")
    goal = profile.get("goal") or "General wellness"
    minutes = int(profile.get("available_minutes") or 35)
    family = _goal_family(goal)
    scheme = _REP_SCHEMES[family]
    difficulty = resolve_difficulty(profile, context)

    candidates = [
        entry for entry in library
        if _available(entry, equipment) and _avoids_limitations(entry, limitations)
    ]
    if not candidates:
        candidates = [entry for entry in _BUILTIN_LIBRARY if _avoids_limitations(entry, limitations)]
    if not candidates:
        candidates = list(_BUILTIN_LIBRARY)

    # One exercise per muscle group, ordered largest movement first, so the
    # session covers the body rather than hammering one area.
    selected: list[dict[str, Any]] = []
    used_groups: set[str] = set()
    for group in _MUSCLE_ORDER:
        for entry in candidates:
            if entry.get("muscle_group") == group and group not in used_groups:
                selected.append(entry)
                used_groups.add(group)
                break

    for entry in candidates:
        if len(selected) >= 5:
            break
        if entry not in selected:
            selected.append(entry)

    # Shorter sessions get fewer movements so the time budget holds.
    limit = 3 if minutes <= 20 else 4 if minutes <= 35 else 5
    selected = selected[:limit]

    sets = scheme["sets"]
    if difficulty.startswith("beginner"):
        sets = max(2, sets - 1)
    if difficulty.endswith("+"):
        sets += 1
    elif difficulty.endswith("-"):
        sets = max(2, sets - 1)

    exercises = [{
        "slug": entry.get("slug"),
        "name": entry.get("name"),
        "muscle_group": entry.get("muscle_group", ""),
        "equipment": entry.get("equipment", "bodyweight"),
        "sets": sets,
        "reps": scheme["reps"],
        "rest_seconds": scheme["rest"],
        "instructions": list(entry.get("instructions") or []),
        "vision_kind": entry.get("vision_kind"),
        "demo_media_url": entry.get("demo_media_url"),
        "cues": list(entry.get("cues") or []),
    } for entry in selected]

    notes = [
        "Prioritize smooth, controlled reps over speed.",
        "Stop a set if your technique breaks down rather than grinding it out.",
    ]
    if limitations:
        notes.insert(0, f"Working around your stated limitation: {limitations}.")

    return {
        "title": f"{goal} session",
        "summary": (
            f"A {minutes}-minute {difficulty} workout built from your goal, equipment, "
            "and recent training history."
        ),
        "duration_minutes": minutes,
        "difficulty": difficulty,
        "goal": goal,
        "warmup": [
            "3 minutes of easy movement to raise your temperature",
            "10 bodyweight good mornings",
            "10 arm circles in each direction",
        ],
        "blocks": [{"name": "Main training block", "focus": goal, "exercises": exercises}],
        "cooldown": [
            "60 seconds of slow nasal breathing",
            "Standing quad stretch, 30 seconds per side",
            "Chest and lat stretch, 30 seconds per side",
        ],
        "coaching_notes": notes,
    }
